src/backend/utils/sort/tuplestore.c

```diff
@@ -1694,21 +1797,83 @@ tuplestore_set_instrument(Tuplestorestate *state,
 /* Extra GPDB functions for sharing tuplestores across processes */
 
 /*
- * tuplestore_make_shared
+ * Initialization of the shared hash table for shared tuplestores.
+ *
+ * XXX: Use MaxBackends to size it, on the assumption that max_connections
+ * will scale accordingly to query complexity. This is quite fuzzy, you could
+ * create a query with tons of tuplestores but only a few
+ * slice, but that ought to be rare enough in practice. This isn't a hard
+ * limit anyway, the hash table will use up any "slop" in shared memory if
+ * needed.
+ */
+#define N_TUPLESTORE_SLOTS() (MaxBackends * 5)
+
+Size
+SharedTuplestoreShmemSize(void)
+{
+	Size		size;
+
+	size = hash_estimate_size(N_TUPLESTORE_SLOTS(), sizeof(TuplestoreSharingState));
+
+	return size;
+}
+
+void
+SharedTuplestoreShmemInit(void)
+{
+	HASHCTL		info;
+
+	info.keysize = NAMEDATALEN;
+	info.entrysize = sizeof(TuplestoreSharingState);
+
+	shared_tuplestores = ShmemInitHash("Shared tuplestores data",
+									   N_TUPLESTORE_SLOTS(),
+									   N_TUPLESTORE_SLOTS(),
+									   &info,
+									   HASH_ELEM);
+}
+
+/*
+ * tuplestore_make_shared_many
  *
  * Make a tuplestore available for sharing later. This must be called
  * immediately after tuplestore_begin_heap().
+ *
+ * The caller must make sure to set transaction-level MemoryContext
+ * and ResourceOwner if this might be called on segments.
+ *
+ * The ntotal parameter sets the number of times the tuplestore can be
+ * opened before it is automatically deleted.
  */
 void
-tuplestore_make_shared(Tuplestorestate *state, SharedFileSet *fileset, const char *filename)
+tuplestore_make_shared_many(Tuplestorestate *state, SharedFileSet *fileset, const char *filename, uint32 ntotal)
 {
 	ResourceOwner oldowner;
+	TuplestoreSharingState *sstate;
+	bool		found;
 
 	state->work_set = workfile_mgr_create_set("SharedTupleStore", filename, true /* hold pin */);
 
 	Assert(state->status == TSS_INMEM);
 	Assert(state->tuples == 0);
 	Assert(state->share_status == TSHARE_NOT_SHARED);
+	/* We only support one shared fileset currently */
+	Assert(fileset == get_shareinput_fileset());
+	/*
```

Comment:
Can we avoid hard-coding the fileset assumption in this low-level module?
To make tuplestore usable in more general contexts.

Maybe store the SharedFileSet pointer inside TuplestoreSharingState
(at the top level).